src/transformStateWithClones.js

```diff
@@ -6,8 +6,32 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+
+function transformStateWithClones(initialState, actions) {
+  const states = [];
+  let currentState = { ...initialState };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      const properties = action.extraData;
+
+      for (const key in properties) {
+        currentState[key] = properties[key];
+      }
+    } else if (action.type === 'removeProperties') {
+      const keysToRemove = action.keysToRemove;
```

Comment:
Consider adding a check to ensure `action.keysToRemove` is an array before iterating over it. This will prevent potential runtime errors if `keysToRemove` is not provided or is not an array.